synthtool/languages/php.py

```diff
@@ -0,0 +1,184 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import contextlib
+import logging
+import os
+from pathlib import Path
+import re
+import subprocess
+import typing
+
+import synthtool as s
+from synthtool.log import logger
+
+
+STAGING_DIR = "owl-bot-staging"
+METADATA_DIR = "GPBMetadata"
+COPYRIGHT_REGEX = re.compile(r"Copyright (\d{4}) Google LLC$", flags=re.MULTILINE)
+OWLBOT_PY_FILENAME = "owlbot.py"
+DEFAULT_COPY_EXCLUDES: typing.List[str] = []
+
+
+@contextlib.contextmanager
+def pushd(d: Path):
+    """Create a context for changing directory.
+
+    When exiting the context, it will go back to the original directory.
+    """
+    original_dir = os.getcwd()
+    os.chdir(d)
+    try:
+        yield
+    finally:
+        os.chdir(original_dir)
+
+
+def _merge(src: str, dest: str, path: Path):
+    """Merge function for the PHP post processor.
+    This should be used for most merges of newly generated and existing files.
+    It preserves copyright year from destination files
+    Args:
+        src: Source file content from gapic
+        dest: Destination file content
+        path: Destination file path
+    Returns:
+        The merged file content.
+    """
+    logger.debug("_merge called for %s", path)
+    m = re.search(COPYRIGHT_REGEX, dest)
+    if m:
+        return re.sub(COPYRIGHT_REGEX, f"Copyright {m.group(1)} Google LLC", src, 1)
+    return src
+
+
+def _find_copy_target(src: Path, version_string: str) -> typing.Optional[Path]:
+    """Returns a directory contains the version subdirectory.
+    """
+    logger.debug("_find_copy_target called with %s and %s", src, version_string)
+    entries = os.scandir(src)
+    if not entries:
+        return None
+    for entry in entries:
+        if entry.path.endswith(version_string):
+            return src
+        if entry.is_dir():
+            return _find_copy_target(Path(entry.path).resolve(), version_string)
+    return None
+
+
+def owlbot_copy_version(
+    src: Path, dest: Path, copy_excludes: typing.Optional[typing.List[str]] = None,
+) -> None:
+    """Copies files from a version subdirectory.
+    """
+    logger.debug("owlbot_copy_version called from %s to %s", src, dest)
+
+    if copy_excludes is None:
+        copy_excludes = DEFAULT_COPY_EXCLUDES
+    # detect the version string for later use
+    entries = os.scandir(src / "src")
+    if not entries:
+        logger.info("there is no src directory to copy")
```

Comment:
Include the directory that was inspected in the log message.  Something like `Found no src directory in ${src}`.

Reply:
Done (in the next commit)

Reply:
I don't see it yet.